contracts/test/integration/L2PlusXNet.t.sol

```diff
@@ -0,0 +1,599 @@
+    mapping(bytes32 => SubnetNode) nodes;
+    mapping(bytes32 => SubnetDefinition) definitions;
+}
+
+struct SubnetDefinition {
+    address gatewayAddr;
+    SubnetID id;
+}
+
+struct SubnetNode {
+    /// @dev The list of child subnet actor address
+    EnumerableSet.AddressSet childActors;
+}
+
+struct SubnetCreationParams {
+    SubnetID parent;
+    Asset supplySource;
+}
+
+library LibSubnetsHierarchy {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SubnetIDHelper for SubnetID;
+
+    function initL1(SubnetsHierarchy storage self, SubnetDefinition memory l1) internal {
+        bytes32 lookupId = l1.id.toHash();
+
+        self.l1NodeLookupKey = lookupId;
+        storeNewNode(self, lookupId, l1.id, l1.gatewayAddr);
+    }
+
+    function getSubnetGateway(SubnetsHierarchy storage self, SubnetID memory id) internal view returns (address) {
+        bytes32 lookupId = id.toHash();
+        return getSubnetGateway(self, lookupId);
+    }
+
+    function getSubnetGateway(SubnetsHierarchy storage self, bytes32 lookupId) internal view returns (address) {
+        require(self.definitions[lookupId].gatewayAddr != address(0), "subnet not found");
+        return self.definitions[lookupId].gatewayAddr;
+    }
+
+    function storeNewNode(
+        SubnetsHierarchy storage self,
+        bytes32 lookupId,
+        SubnetID memory id,
+        address gateway
+    ) internal {
+        self.definitions[lookupId].gatewayAddr = gateway;
+        self.definitions[lookupId].id = id;
+    }
+
+    function linkNewChild(SubnetsHierarchy storage self, bytes32 lookupId, address subnetActor) internal {
+        self.nodes[lookupId].childActors.add(subnetActor);
+    }
+
+    function registerNewSubnet(
+        SubnetsHierarchy storage self,
+        SubnetID memory parent,
+        address subnetActor,
+        address gateway
+    ) internal returns (SubnetID memory) {
+        // ensure parent exists
+        getSubnetGateway(self, parent);
+
+        SubnetID memory child = SubnetIDHelper.createSubnetId(parent, subnetActor);
+
+        bytes32 lookupId = child.toHash();
+
+        self.actorToLookupKeys[subnetActor] = lookupId;
+        storeNewNode(self, lookupId, child, gateway);
+        linkNewChild(self, lookupId, subnetActor);
+
+        return child;
+    }
+
+    function getGatewayBySubnetActor(
+        SubnetsHierarchy storage self,
+        address subnetActor
+    ) internal view returns (address) {
+        bytes32 lookupId = self.actorToLookupKeys[subnetActor];
+        return getSubnetGateway(self, lookupId);
+    }
+}
+
+interface IGateway {
+    function sendContractXnetMessage(
+        IpcEnvelope memory envelope
+    ) external payable returns (IpcEnvelope memory committed);
+
+    function fund(SubnetID calldata subnetId, FvmAddress calldata to) external payable;
+
+    function fundWithToken(SubnetID calldata subnetId, FvmAddress calldata to, uint256 amount) external;
+}
```

Comment:
Do we need this local interface definition here? Can't we import the canonical one from `interfaces/`?